src/domain/contracts/decoders/abi-decoder.helper.ts

```diff
@@ -1,78 +1,165 @@
+
+type FunctionDataDecoders<TAbi extends Abi> = {
+  [key in FunctionDataDecoder<TAbi>]: (
+    data: Hex,
+  ) => ContractFunctionArgs<TAbi, AbiStateMutability, key> | null;
+};
+
+export function _generateFunctionDataDecoders<TAbi extends Abi>(
+  abi: TAbi,
+): FunctionDataDecoders<TAbi> {
+  const createFunctionDataDecoder = (name: FunctionDataDecoder<TAbi>) => {
+    return (
+      data: `0x${string}`,
+    ): ContractFunctionArgs<
+      TAbi,
+      AbiStateMutability,
+      ContractFunctionName<TAbi>
+    > | null => {
+      try {
+        const decoded = decodeFunctionData({
+          data,
+          abi,
+        });
+
+        if (decoded.functionName !== name) {
+          return null;
+        }
+
+        return decoded.args as ContractFunctionArgs<
+          TAbi,
+          AbiStateMutability,
+          ContractFunctionName<TAbi>
+        >;
+      } catch {
+        return null;
+      }
     };
+  };
+
+  const functionDataDecoders = {} as FunctionDataDecoders<TAbi>;
+
+  for (const item of abi) {
+    if (item.type === 'function') {
+      const name = item.name as FunctionDataDecoder<TAbi>;
+      functionDataDecoders[name] = createFunctionDataDecoder(name);
+    }
   }
 
-  return helpers;
+  return functionDataDecoders;
 }
 
-export abstract class AbiDecoder<TAbi extends Abi> {
-  readonly helpers: Helpers<TAbi>;
+// Event log decoders
 
-  protected constructor(private readonly abi: TAbi) {
-    this.helpers = _generateHelpers(abi);
-  }
+type EventLogDecoder<TAbi extends Abi> = ContractEventName<TAbi>;
 
-  // Use inferred types from viem
-  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-  decodeEventLog<
-    const abi extends TAbi,
-    eventName extends ContractEventName<abi> | undefined = undefined,
-    topics extends Hex[] = Hex[],
-    data extends Hex | undefined = undefined,
-    strict extends boolean = true,
-  >(
-    args: Omit<
-      DecodeEventLogParameters<abi, eventName, topics, data, strict>,
-      'abi'
-    >,
-  ) {
-    return _decodeEventLog({
-      ...args,
-      abi: this.abi,
-    });
-  }
+type EventLogDecoders<TAbi extends Abi> = {
+  [key in EventLogDecoder<TAbi>]: (args: {
+    data: `0x${string}`;
+    topics: [signature: `0x${string}`, ...args: Array<`0x${string}`>] | [];
+  }) => ContractEventArgsFromTopics<TAbi, key> | null;
+};
+
+export function _generateEventLogDecoders<TAbi extends Abi>(
+  abi: TAbi,
+): EventLogDecoders<TAbi> {
+  const createEventLogDecoder = (name: EventLogDecoder<TAbi>) => {
+    return (args: {
+      data: `0x${string}`;
+      topics: [signature: `0x${string}`, ...args: Array<`0x${string}`>] | [];
+    }): ContractEventArgsFromTopics<TAbi, ContractEventName<TAbi>> | null => {
+      try {
+        const decoded = decodeEventLog({
+          abi,
+          data: args.data,
+          topics: args.topics,
+        });
+
+        if (decoded.eventName !== name) {
+          return null;
```

Comment:
It can happen when if the function data is of a function in the ABI, but not that one expected. Changed it to log/throw in 36b45c17e2e254e6365f838b1b14e566de930810.